Use align-content when calculating the static position of absolutely-positioned flexbox children.

rreno opened the pull request:
#### 62e012dfc64a13172b55d9115d77119ec8dbe794
<pre>
Use align-content when calculating the static position of absolutely-positioned flexbox children.
<a href="https://bugs.webkit.org/show_bug.cgi?id=221472">https://bugs.webkit.org/show_bug.cgi?id=221472</a>
rdar://74278896

Reviewed by Sergio Villar Senin.

Absolutely-positioned children of multi-line flex containers were not taking into account
the value of align-content. Since abspos children are statically positioned as if they
were the only flex item, we must compute the alignment of the hypothetical flex line that
they would be on.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild const):
    Added a lambda for computing the alignment offset of the hypothetical flex line that
    an abspos child of a flex container lives on. We don&apos;t fully implement the spec
    at this time since there is a webcompat issue for fallback behavior for
    align-content: stretch. We maintain our current behavior and compatibility with all other
    engines and we compute position as if the hypothetical flex line stretches to fill the container.

(WebCore::RenderFlexibleBox::availableAlignmentSpaceForChild const):
(WebCore::RenderFlexibleBox::availableAlignmentSpaceForChild): Deleted.
    Made into a const member function.
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild): Deleted.
    Made into a const member function.
* Source/WebCore/rendering/RenderFlexibleBox.h:

LayoutTests:
    Rebaseline css/css-flexbox/abspos tests.
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-rtl-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-rtl-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-vertWM-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-vertWM-002-expected.txt:

Canonical link: <a href="https://commits.webkit.org/253389@main">https://commits.webkit.org/253389@main</a>
</pre>

svillar commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1731,10 +1731,48 @@ LayoutUnit RenderFlexibleBox::staticMainAxisPositionForPositionedChild(const Ren
+    // since that behavior is the same as in single-line flex containers. Note, this is not what the spec
+    // calls for when availableSpace is negative but it's what browsers have been shipping.
+    // Until the spec wording changes to match behavior then we are ignoring the spec for
+    // that one case in the name of web compatibility.
```
Do we have a bug for this? If not, it'd be nice to have one in WebKit, so that future spec/interop discussions could refer to the different bug trackers.

You could even add there a complete description of the issue there and just include the link to the issue here mentioning that we don't follow the spec in this particular case.

rreno replied:
That's a great point. I wrote up a CSSWG issue but didn't make a corresponding bug in our bug tracker. I'll do that and then add the links here.

svillar commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1731,10 +1731,48 @@ LayoutUnit RenderFlexibleBox::staticMainAxisPositionForPositionedChild(const Ren
+    // Until the spec wording changes to match behavior then we are ignoring the spec for
+    // that one case in the name of web compatibility.
+    // https://drafts.csswg.org/css-flexbox-1/#align-content-property
+    const auto alignContentOffset = [&] {
```
I prefer to capture only what's needed but not an strong opinion in this case.

rreno replied:
I also prefer that but in this case we'd be capturing everything except `child`. I chose to capture the unneeded reference instead of making a long list of captures.

svillar reviewed (approved):
Looks great! I've pretty bad memories of static positioning + flexbox :smile:

rreno committed:
Use align-content when calculating the static position of absolutely-positioned flexbox children.
https://bugs.webkit.org/show_bug.cgi?id=221472
rdar://74278896

Reviewed by Sergio Villar Senin.

Absolutely-positioned children of multi-line flex containers were not taking into account
the value of align-content. Since abspos children are statically positioned as if they
were the only flex item, we must compute the alignment of the hypothetical flex line that
they would be on.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild const):
    Added a lambda for computing the alignment offset of the hypothetical flex line that
    an abspos child of a flex container lives on. We don't fully implement the spec
    at this time since there is a webcompat issue for fallback behavior for
    align-content: stretch. We maintain our current behavior and compatibility with all other
    engines and we compute position as if the hypothetical flex line stretches to fill the container.

(WebCore::RenderFlexibleBox::availableAlignmentSpaceForChild const):
(WebCore::RenderFlexibleBox::availableAlignmentSpaceForChild): Deleted.
    Made into a const member function.
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild): Deleted.
    Made into a const member function.
* Source/WebCore/rendering/RenderFlexibleBox.h:

LayoutTests:
    Rebaseline css/css-flexbox/abspos tests.
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-rtl-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-rtl-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-vertWM-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/abspos/flex-abspos-staticpos-align-content-vertWM-002-expected.txt:

Canonical link: https://commits.webkit.org/253389@main
